<a href="https://colab.research.google.com/github/slp22/data-engineering-project/blob/main/engineering_monkeypox_mvp.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### Data Engineering | MVP

# Monkeypox Tweets

## Imports

In [1]:
import json
import logging
import sqlite3
import matplotlib.pyplot as plt
import numpy as np
import os, shutil, itertools
import pandas as pd
import pathlib as Path
import pickle
import PIL
import random
import seaborn as sns
import sklearn as sk
import warnings
import zipfile

from sqlite3 import connect


## py/spark install



In [2]:
# # https://towardsdatascience.com/pyspark-on-google-colab-101-d31830b238be
# # https://www.analyticsvidhya.com/blog/2020/11/a-must-read-guide-on-how-to-work-with-pyspark-on-google-colab-for-data-scientists/

!apt-get install openjdk-8-jdk-headless -qq > /dev/null

In [3]:
!wget -q https://archive.apache.org/dist/spark/spark-3.0.0/spark-3.0.0-bin-hadoop3.2.tgz

In [4]:
!tar -xf spark-3.0.0-bin-hadoop3.2.tgz

In [5]:
!pip install -q findspark

In [6]:
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-3.0.0-bin-hadoop3.2"

In [7]:
import findspark
findspark.init()

In [8]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql.functions import split, explode, col, lower
from pyspark.sql.types import StructType,StructField, StringType, IntegerType


## 1 | Research Design


* **Research Question:** 
* **Impact Hypothesis:** 
* **Data source:** 

* **Data Dictionary:**


## 2 | Data Ingestion

#### [Monkeypox Tweets](https://www.kaggle.com/datasets/aneeshtickoo/tweets-on-monkeypox)

In [9]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [10]:
# copy kaggle key from Google Drive
shutil.copyfile('/content/drive/MyDrive/kaggle.json', '/content/kaggle.json')

'/content/kaggle.json'

In [11]:
#confirm kaggle key is in dir
os.listdir()

['.config',
 'monkeypox.csv',
 'spark-3.0.0-bin-hadoop3.2.tgz.1',
 'spark-3.0.0-bin-hadoop3.2',
 'spark-3.3.0-bin-hadoop3.tgz.1',
 'tweets.pkl',
 'spark-3.0.0-bin-hadoop3.2.tgz',
 'monkeypox.db',
 'tweets-on-monkeypox.zip',
 'kaggle.json',
 'drive',
 'spark-3.3.0-bin-hadoop3.tgz',
 'tweets.csv',
 'sample_data']

In [12]:
with open('kaggle.json') as json_file:
    kaggle = json.load(json_file)

In [13]:
# assign to directory 
os.environ['KAGGLE_CONFIG_DIR'] = "/content"

In [14]:
# download dataset from kaggle
# 'chmod 600 /content/kaggle.json'
! kaggle datasets download -d aneeshtickoo/tweets-on-monkeypox

tweets-on-monkeypox.zip: Skipping, found more recently modified local copy (use --force to force download)


In [15]:
# unzip kaggle file
zip_ref = zipfile.ZipFile('tweets-on-monkeypox.zip', 'r') 
zip_ref.extractall('/content')
zip_ref.close()

## 3 | Exploratory Data Analysis

### First glance

In [16]:
df = pd.read_csv('/content/monkeypox.csv')

In [17]:
df.head(2)

id      conversation_id  \
0  1555815462201872385  1555815462201872385   
1  1555815458602831872  1555815458602831872   

                                created_at        date      time  timezone  \
0  2022-08-06 12:48:06 India Standard Time  2022-08-06  12:48:06       530   
1  2022-08-06 12:48:05 India Standard Time  2022-08-06  12:48:05       530   

              user_id      username                           name place  ...  \
0  820113517613154304  thetenth2022                       TheTenth   NaN  ...   
1           196518052     ashemedai  Jeroen Ruigrok van der Werven   NaN  ...   

  geo source user_rt_id user_rt retweet_id  reply_to  retweet_date  translate  \
0 NaN    NaN        NaN     NaN        NaN        []           NaN        NaN   
1 NaN    NaN        NaN     NaN        NaN        []           NaN        NaN   

  trans_src trans_dest  
0       NaN        NaN  
1       NaN        NaN  

[2 rows x 36 columns]

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6859 entries, 0 to 6858
Data columns (total 36 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               6859 non-null   int64  
 1   conversation_id  6859 non-null   int64  
 2   created_at       6859 non-null   object 
 3   date             6859 non-null   object 
 4   time             6859 non-null   object 
 5   timezone         6859 non-null   int64  
 6   user_id          6859 non-null   int64  
 7   username         6859 non-null   object 
 8   name             6859 non-null   object 
 9   place            2 non-null      object 
 10  tweet            6859 non-null   object 
 11  language         6859 non-null   object 
 12  mentions         6859 non-null   object 
 13  urls             6859 non-null   object 
 14  photos           6859 non-null   object 
 15  replies_count    6859 non-null   int64  
 16  retweets_count   6859 non-null   int64  
 17  likes_count   

In [19]:
df.describe()

id  conversation_id  timezone       user_id  replies_count  \
count  6.859000e+03     6.859000e+03    6859.0  6.859000e+03    6859.000000   
mean   1.555764e+18     1.555192e+18     530.0  7.224884e+17       0.399038   
std    2.541498e+13     1.168003e+16       0.0  6.620987e+17       1.789475   
min    1.555725e+18     8.510359e+17     530.0  3.989300e+04       0.000000   
25%    1.555742e+18     1.555720e+18     530.0  4.906582e+08       0.000000   
50%    1.555760e+18     1.555746e+18     530.0  8.977771e+17       0.000000   
75%    1.555784e+18     1.555774e+18     530.0  1.395422e+18       0.000000   
max    1.555815e+18     1.555815e+18     530.0  1.555798e+18      68.000000   

       retweets_count  likes_count        video  near  geo  source  \
count     6859.000000  6859.000000  6859.000000   0.0  0.0     0.0   
mean         0.691938     3.053944     0.095932   NaN  NaN     NaN   
std         11.879167    40.567536     0.294520   NaN  NaN     NaN   
min          0.000000     0.000000     0.000000   NaN  NaN     NaN   
25%          0.000000     0.000000     0.000000   NaN  NaN     NaN   
50%          0.000000     0.000000     0.000000   NaN  NaN     NaN   
75%          0.000000     1.000000     0.000000   NaN  NaN     NaN   
max        740.000000  2180.000000     1.000000   NaN  NaN     NaN   

       user_rt_id  user_rt  retweet_id  retweet_date  translate  trans_src  \
count         0.0      0.0         0.0           0.0        0.0        0.0   
mean          NaN      NaN         NaN           NaN        NaN        NaN   
std           NaN      NaN         NaN           NaN        NaN        NaN   
min           NaN      NaN         NaN           NaN        NaN        NaN   
25%           NaN      NaN         NaN           NaN        NaN        NaN   
50%           NaN      NaN         NaN           NaN        NaN        NaN   
75%           NaN      NaN         NaN           NaN        NaN        NaN   
max           NaN      NaN         NaN           NaN        NaN        NaN   

       trans_dest  
count         0.0  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

In [20]:
cols_list = list(df.columns)
cols_list

['id',
 'conversation_id',
 'created_at',
 'date',
 'time',
 'timezone',
 'user_id',
 'username',
 'name',
 'place',
 'tweet',
 'language',
 'mentions',
 'urls',
 'photos',
 'replies_count',
 'retweets_count',
 'likes_count',
 'hashtags',
 'cashtags',
 'link',
 'retweet',
 'quote_url',
 'video',
 'thumbnail',
 'near',
 'geo',
 'source',
 'user_rt_id',
 'user_rt',
 'retweet_id',
 'reply_to',
 'retweet_date',
 'translate',
 'trans_src',
 'trans_dest']

In [21]:
df

id      conversation_id  \
0     1555815462201872385  1555815462201872385   
1     1555815458602831872  1555815458602831872   
2     1555815444359135235  1555815444359135235   
3     1555815437409157120  1555815437409157120   
4     1555815424666869761  1555595696639774722   
...                   ...                  ...   
6854  1555725304659935232  1555718508813570048   
6855  1555725295432245248  1555725295432245248   
6856  1555725294685900801  1555655517850128384   
6857  1555725290055385089  1555725290055385089   
6858  1555725288201494529  1555725288201494529   

                                   created_at        date      time  timezone  \
0     2022-08-06 12:48:06 India Standard Time  2022-08-06  12:48:06       530   
1     2022-08-06 12:48:05 India Standard Time  2022-08-06  12:48:05       530   
2     2022-08-06 12:48:02 India Standard Time  2022-08-06  12:48:02       530   
3     2022-08-06 12:48:00 India Standard Time  2022-08-06  12:48:00       530   
4     2022-08-06 12:47:57 India Standard Time  2022-08-06  12:47:57       530   
...                                       ...         ...       ...       ...   
6854  2022-08-06 06:49:51 India Standard Time  2022-08-06  06:49:51       530   
6855  2022-08-06 06:49:49 India Standard Time  2022-08-06  06:49:49       530   
6856  2022-08-06 06:49:48 India Standard Time  2022-08-06  06:49:48       530   
6857  2022-08-06 06:49:47 India Standard Time  2022-08-06  06:49:47       530   
6858  2022-08-06 06:49:47 India Standard Time  2022-08-06  06:49:47       530   

                  user_id         username                           name  \
0      820113517613154304     thetenth2022                       TheTenth   
1               196518052        ashemedai  Jeroen Ruigrok van der Werven   
2               600467609  democracymotion            Democracy In Motion   
3     1431886953998258177      thegoogle93                     The Google   
4     1506745578159591428   bufflosouljah1                  Bufflosouljah   
...                   ...              ...                            ...   
6854  1498464506384732162     arkcowgirl62                  ValTheCowgirl   
6855             38338051     speehanagram                        Spencer   
6856  1548821540577464322   pritzkertoilet         Rev. Strychnine Twitch   
6857            298690857     thumbressler                      Mike Reux   
6858           4069820847  nurse4allpeople              (((Love4Planet)))   

     place  ... geo source user_rt_id user_rt retweet_id  \
0      NaN  ... NaN    NaN        NaN     NaN        NaN   
1      NaN  ... NaN    NaN        NaN     NaN        NaN   
2      NaN  ... NaN    NaN        NaN     NaN        NaN   
3      NaN  ... NaN    NaN        NaN     NaN        NaN   
4      NaN  ... NaN    NaN        NaN     NaN        NaN   
...    ...  ...  ..    ...        ...     ...        ...   
6854   NaN  ... NaN    NaN        NaN     NaN        NaN   
6855   NaN  ... NaN    NaN        NaN     NaN        NaN   
6856   NaN  ... NaN    NaN        NaN     NaN        NaN   
6857   NaN  ... NaN    NaN        NaN     NaN        NaN   
6858   NaN  ... NaN    NaN        NaN     NaN        NaN   

                                               reply_to  retweet_date  \
0                                                    []           NaN   
1                                                    []           NaN   
2                                                    []           NaN   
3                                                    []           NaN   
4     [{'screen_name': 'Natrone86', 'name': 'Weed Al...           NaN   
...                                                 ...           ...   
6854  [{'screen_name': 'hurtmeknots', 'name': "Sambe...           NaN   
6855                                                 []           NaN   
6856  [{'screen_name': 'politvidchannel', 'name': 'P...           NaN   
6857                                                 []           NaN   


* RangeIndex: 6859 entries, 0 to 6858
* Data columns (total 36 columns)

### word count

In [22]:
# type(df)
# df.head(2)
pandas_df = df[['date','tweet']]
pandas_df.head(2)

date                                              tweet
0  2022-08-06  So has anyone begun to compile 'here's the sta...
1  2022-08-06  Getting some groceries, topic shifted to covid...

In [23]:
spark = SparkSession.builder\
        .master("local")\
        .appName("Colab")\
        .config('spark.ui.port', '4050')\
        .getOrCreate()

In [24]:
df_schema = StructType([StructField("date", StringType(), True),
                        StructField("tweet", StringType(), True)])


In [25]:
# tweet_df = spark.read.csv('/content/tweets.csv', inferSchema=True, header='true')

spark_df = spark.createDataFrame(pandas_df, schema=df_schema)
spark_df.printSchema()
spark_df.show()

root
 |-- date: string (nullable = true)
 |-- tweet: string (nullable = true)

+----------+--------------------+
|      date|               tweet|
+----------+--------------------+
|2022-08-06|So has anyone beg...|
|2022-08-06|Getting some groc...|
|2022-08-06|"Illinois Childre...|
|2022-08-06|Illinois daycare ...|
|2022-08-06|@Natrone86 @Junot...|
|2022-08-06|"CDC Recommends L...|
|2022-08-06|1,058,637 America...|
|2022-08-06|MF didn’t know sh...|
|2022-08-06|Three people of A...|
|2022-08-06|MonkeyPox Case: ರ...|
|2022-08-06|#Monkeypox Fallza...|
|2022-08-06|RT-PCR detection ...|
|2022-08-06|@BigBadDenis When...|
|2022-08-06|News Outlook TOP ...|
|2022-08-06|CDC Issues Eye-Op...|
|2022-08-06|United States dec...|
|2022-08-06|@FoxNews They are...|
|2022-08-06|@PetiteNicoco No....|
|2022-08-06|Idc if it’s Monke...|
|2022-08-06|In the technical ...|
+----------+--------------------+
only showing top 20 rows



In [26]:
spark_df = spark_df.withColumn('tweet', 
                               explode(split(lower(col('tweet')), '\s')))

In [27]:
(spark_df.groupBy('tweet')
  .count()
  .orderBy('count', ascending=False)
  .show(10))

+---------+-----+
|    tweet|count|
+---------+-----+
|         | 4898|
|      the| 4351|
|monkeypox| 3230|
|       to| 2957|
|        a| 2400|
|      and| 2344|
|   monkey| 2151|
|       is| 2014|
|       of| 1901|
|        i| 1721|
+---------+-----+
only showing top 10 rows



#❓Count specific words using PySpark?

In [28]:
(spark_df)

DataFrame[date: string, tweet: string]

### Corpus `tweets`

In [29]:
tweets = df[['language','date','username','hashtags','tweet']]
tweets.head(2)

language        date      username hashtags  \
0       en  2022-08-06  thetenth2022       []   
1       en  2022-08-06     ashemedai       []   

                                               tweet  
0  So has anyone begun to compile 'here's the sta...  
1  Getting some groceries, topic shifted to covid...

In [30]:
tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6859 entries, 0 to 6858
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   language  6859 non-null   object
 1   date      6859 non-null   object
 2   username  6859 non-null   object
 3   hashtags  6859 non-null   object
 4   tweet     6859 non-null   object
dtypes: object(5)
memory usage: 268.1+ KB


In [31]:
tweets['username'].nunique()

6028

In [32]:
# How many languages? 
tweets['language'].unique()

array(['en', 'kn', 'de', 'in', 'tl', 'fr', 'pt', 'te', 'tr', 'it', 'qme',
       'bn', 'qht', 'pl', 'es', 'el', 'nl', 'cy', 'ta', 'hi', 'sv', 'ja',
       'th', 'mr', 'et', 'gu', 'da', 'ro', 'ml', 'zxx', 'und', 'pa', 'ur',
       'ko', 'am', 'fi', 'zh', 'lt', 'hu', 'ru', 'ar', 'si'], dtype=object)

In [33]:
# How many tweets in English?
print('English entries:', (tweets[tweets["language"] == 'en'].count())['language'])

# How many tweets in other languages?
print('Spanish entries:', (tweets[tweets["language"] == 'es'].count())['language'])
print('Italian entries:', (tweets[tweets["language"] == 'it'].count())['language'])

English entries: 6410
Spanish entries: 51
Italian entries: 10


In [34]:
# Keep only English languge tweets
tweets = tweets[(tweets['language'] == 'en')]

In [35]:
tweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6410 entries, 0 to 6858
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   language  6410 non-null   object
 1   date      6410 non-null   object
 2   username  6410 non-null   object
 3   hashtags  6410 non-null   object
 4   tweet     6410 non-null   object
dtypes: object(5)
memory usage: 300.5+ KB


In [36]:
tweets.head(2)

language        date      username hashtags  \
0       en  2022-08-06  thetenth2022       []   
1       en  2022-08-06     ashemedai       []   

                                               tweet  
0  So has anyone begun to compile 'here's the sta...  
1  Getting some groceries, topic shifted to covid...

In [37]:
# Drop language column since we have only English tweets
tweets = tweets.drop(columns=['language'])

In [38]:
tweets.head(10)

date         username hashtags  \
0   2022-08-06     thetenth2022       []   
1   2022-08-06        ashemedai       []   
2   2022-08-06  democracymotion       []   
3   2022-08-06      thegoogle93       []   
4   2022-08-06   bufflosouljah1       []   
5   2022-08-06  democracymotion       []   
6   2022-08-06       salemjakes       []   
7   2022-08-06      fineassbrei       []   
8   2022-08-06         xenohadi       []   
11  2022-08-06   geopoliticsind       []   

                                                tweet  
0   So has anyone begun to compile 'here's the sta...  
1   Getting some groceries, topic shifted to covid...  
2   "Illinois Children's Daycare Worker Tests Posi...  
3   Illinois daycare worker tests positive for mon...  
4   @Natrone86 @JunotIsrael @elcavaqueen @thechicc...  
5   "CDC Recommends Limiting Sex Partners to Avoid...  
6   1,058,637 Americans have died in the U.S. Covi...  
7   MF didn’t know shit about monkey pox in June c...  
8   Three people of African origin still admitted ...  
11  RT-PCR detection kit for human monkeypox virus...

In [39]:
# Save corpus
tweets.to_pickle('/content/tweets.pkl')
tweets.to_csv(r'/content/tweets.csv', index=False)

# 4 | Storage

#### SQL Database `monkeypox.db`

##### helper functions

In [40]:
# https://towardsdatascience.com/have-a-sql-interview-coming-up-ace-it-using-google-colab-6d3c0ffb29dc

def pd_to_sqlDB(input_df: pd.DataFrame,
                table_name: str,
                db_name: str = 'default.db') -> None:

    # # Setup local logging
    # logging.basicConfig(level=logging.INFO,
    #                     format='%(asctime)s %(levelname)s: %(message)s',
    #                     datefmt='%Y-%m-%d %H:%M:%S')

    # Find columns in the dataframe
    cols = input_df.columns
    cols_string = ','.join(cols)
    val_wildcard_string = ','.join(['?'] * len(cols))

    # Connect to a DB file if it exists, else create a new file
    con = sqlite3.connect(db_name)
    cur = con.cursor()
    # logging.info(f'SQL DB {db_name} created')

    # Create table
    sql_string = f"""CREATE TABLE {table_name} ({cols_string});"""
    cur.execute(sql_string)
    # logging.info(f'SQL Table {table_name} created with {len(cols)} columns')

    # Upload df
    rows_to_upload = input_df.to_dict(orient='split')['data']
    sql_string = f"""INSERT INTO {table_name} ({cols_string}) VALUES ({val_wildcard_string});"""    
    cur.executemany(sql_string, rows_to_upload)
    # logging.info(f'{len(rows_to_upload)} rows uploaded to {table_name}')
  
    # Commit the changes and close the connection
    con.commit()
    con.close()

In [41]:
#  https://towardsdatascience.com/have-a-sql-interview-coming-up-ace-it-using-google-colab-6d3c0ffb29dc

def sql_query_to_pd(sql_query_string: str, db_name: str ='mpox.db') -> pd.DataFrame:
    
    # Connect to the SQL DB
    con = sqlite3.connect(db_name)

    # Execute the SQL query
    cursor = con.execute(sql_query_string)

    # Fetch the data and column names
    result_data = cursor.fetchall()
    cols = [description[0] for description in cursor.description]

    # Close the connection
    con.close()

    # Return as df
    return pd.DataFrame(result_data, columns=cols)

##### sql_to_df

In [43]:
# https://towardsdatascience.com/have-a-sql-interview-coming-up-ace-it-using-google-colab-6d3c0ffb29dc

# Read  csv as df
input_df = pd.read_csv('/content/tweets.csv')

# Upload df to a SQL table
pd_to_sqlDB(input_df,
            table_name='tweets',
            db_name='monkeypox.db')

# Write SQL query in a string variable
sql_query_string = """
    SELECT *
    FROM tweets
"""
# Exectue  SQL query
corpus = sql_query_to_pd(sql_query_string, db_name='monkeypox.db')
corpus

date         username                         hashtags  \
0     2022-08-06     thetenth2022                               []   
1     2022-08-06        ashemedai                               []   
2     2022-08-06  democracymotion                               []   
3     2022-08-06      thegoogle93                               []   
4     2022-08-06   bufflosouljah1                               []   
...          ...              ...                              ...   
6405  2022-08-06     arkcowgirl62                               []   
6406  2022-08-06     speehanagram          ['monkeypox', 'besafe']   
6407  2022-08-06   pritzkertoilet                               []   
6408  2022-08-06     thumbressler                    ['monkeypox']   
6409  2022-08-06  nurse4allpeople  ['moscowmitch', 'donothinggqp']   

                                                  tweet  
0     So has anyone begun to compile 'here's the sta...  
1     Getting some groceries, topic shifted to covid...  
2     "Illinois Children's Daycare Worker Tests Posi...  
3     Illinois daycare worker tests positive for mon...  
4     @Natrone86 @JunotIsrael @elcavaqueen @thechicc...  
...                                                 ...  
6405  @hurtmeknots With Monkeypox on the rise, maybe...  
6406  @FITNESSSF can you start filling up the disinf...  
6407     @politvidchannel 71% already have monkey pox .  
6408                          RT if you have #monkeypox  
6409  @POTUS @VP 🥰 Dems 🇺🇸💙 are having the best jobs...  

[6410 rows x 4 columns]

* **database = `monkeypox.db`**
* **table_1 = `corpus` (6410 rows × 4 columns)**

In [44]:
# copy corpus csv 
shutil.copyfile('/content/tweets.csv', '/content/drive/MyDrive/tweets.csv')

'/content/drive/MyDrive/tweets.csv'

# 5 | Processing

* natural language processing, topic modeling (LDA)  
* symptom trends across time + moving average (% or |n|) 

# 6 | Deployment

In [45]:
# streamlit demo: dataframe with countries > adapt to states
# http://localhost:8501/DataFrame_Demo

In [46]:
# dataset with locations to map cases worldwide
# https://www.kaggle.com/datasets/deepcontractor/monkeypox-dataset-daily-updated?select=Monkey_Pox_Cases_Worldwide.csv

# US
# https://www.cdc.gov/poxvirus/monkeypox/response/2022/us-map.html
# https://www.cdc.gov/9cafd3c7-688d-434c-8f1f-7fb329ff618f #csv file

In [47]:
!pip install streamlit

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [48]:
!streamlit hello

2022-08-28 17:03:50.046 INFO    numexpr.utils: NumExpr defaulting to 2 threads.

  Welcome to Streamlit. Check out our demo in your browser.

  Network URL: http://172.28.0.2:8501
  External URL: http://35.196.24.7:8501

  Ready to create your own Python apps super quickly?
  Head over to https://docs.streamlit.io

  May you create awesome apps!


  Stopping...
^C


# 7 | Testing/Robustness

[Python schedule](https://schedule.readthedocs.io/en/stable/examples.html#run-a-job-every-x-minute)